src/include/duckdb/main/capi/capi_internal.hpp

```diff
@@ -28,8 +29,12 @@
 
 namespace duckdb {
 
-struct DatabaseData {
-	unique_ptr<DuckDB> database;
+struct DBInstanceCacheWrapper {
+	unique_ptr<DBInstanceCache> instance_cache;
+};
+
+struct DatabaseWrapper {
+	shared_ptr<DuckDB> database;
```

Comment:
Turning this into a `shared_ptr` instead of the previous `unique_ptr` is probably the most relevant change. This allows for the following, but I am unsure if I missed some side effects of that change.
```cpp
		if (cache) {
 			wrapper->database = cache->instance_cache->GetOrCreateInstance(path, *db_config, true);
 		} else {
 			wrapper->database = duckdb::make_shared_ptr<DuckDB>(path, db_config);
 		}
```